File: 0121_BestTimeToBuyAndSellStock.java
class Solution {
    public int maxProfit(int[] prices) {
        int res = 0;
        int left = 0, right = 0;

        while (right != prices.length) {
            res = Math.max(res, prices[right] - prices[left]);
            if (left == right)
                right++;
            else if (prices[right] < prices[left])
                left++;
            else
                right++;
        }

        return res;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] prices = {7,6,4,3,1};
        System.out.println(sol.maxProfit(prices));
    }
}

File: 0121_best_time_to_buy_and_sell_stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        i, j = 0, 0
        m = 0
        while j < len(prices):
            m = max(m, prices[j] - prices[i])
            if i == j:
                j += 1
            elif prices[j] < prices[i]:
                i += 1
            else:
                j += 1
        return m
